Moved cluster topology information to the topology overview section (#7144)
ydb-platform/ydb: ydb/docs/en/core/concepts/topology.md

* [Database nodes](./glossary.md#database-node) enable query execution, transaction coordination, and other data control functionality.

Both storage and database [nodes](./glossary.md#node) are server processes running an executable called `ydbd`. Storage and database nodes run on one or more physical servers, either directly in the operating system of each physical server, or in the container environment or virtual machines.

To ensure the fault tolerance of a {{ ydb-short-name }} cluster, properly configure the [distributed storage](glossary.md#distributed-storage). Select the appropriate [operating mode](#cluster-config) for your cluster according to the workload requirements, and adjust for [additional fault tolerance requirements](#fault-tolerance).

## Cluster operating modes {#cluster-config}

Cluster topology is based on the chosen operating mode, which is determined by the fault tolerance requirements. A fail model of {{ ydb-short-name }} is based on concepts such as [fail domain](./glossary.md#fail-domain) and [fail realm](./glossary.md#fail-realm).

The following operating modes are available:

> * `none` - There is no redundancy. Any hardware failure causes the storage to become unavailable. This mode is only recommended for functional testing.
> * `block-4-2` - [Erasure coding](https://en.wikipedia.org/wiki/Erasure_code) with two blocks of redundancy added to the four blocks of source data is applied. Storage nodes are placed in at least 8 failure domains (usually racks). The storage pool is available if any two domains fail, continuing to record all 6 data parts in the remaining domains. This mode is recommended for storage pools within a single availability zone (usually a data processing center).
> * `mirror-3-dc` - Data is replicated to 3 availability zones using 3 failure domains (usually racks) within each zone. The storage pool is available if one availability zone and one failure domain fail in the remaining zones. This mode is recommended for multi-data center installations.
> * `mirror-3-dc-3-nodes` - A simplified version of `mirror-3-dc`. This mode requires at least 3 servers with 3 disks each. Each server must be located in an independent data center in order to provide the best fault tolerance. Health in this mode is maintained if no more than 1 node fails. This mode is only recommended for functional testing.

Reviewer: ```suggestion
- `none`. There is no redundancy. Any hardware failure causes data to become unavailable or permanently lost. This mode is only recommended for development and functional testing.
- `block-4-2`. [Erasure coding](https://en.wikipedia.org/wiki/Erasure_code) is applied with two blocks of redundancy added to the four blocks of source data. Storage nodes are placed in at least 8 failure domains (usually racks). A [storage pool](./glossary.md#storage-pool) remains available if any two domains fail, continuing to record all 6 data parts in the remaining domains. This mode is recommended for clusters deployed within a single availability zone or data center.
- `mirror-3-dc`. Data is replicated to 3 availability zones using 3 failure domains (usually racks) within each zone. {{ ydb-short-name }} cluster remains available even if one availability zone completely fails; additionally, one failure domain in the remaining zones can fail at the same time. This mode is recommended for multi-datacenter clusters with high availability requirements.
- `mirror-3-dc-3-nodes`. A simplified version of `mirror-3-dc`. This mode requires at least 3 servers with 3 disks each. Each server must be located in an independent data center to provide reasonable fault tolerance. System health in this mode is maintained if no more than one server fails. This mode is only recommended for functional testing or building prototypes.
```

{% note info %}

Node failure means both its total and partial unavailability, for example, failure of a single disk on a node.

{% endnote %}

> The table below describes the requirements and fault tolerance levels for different operating modes:
> 
> | Mode | Storage<br>volume multiplier | Minimum<br>number<br>of nodes | Fail<br>domain | Fail<br>realm | Number of<br>data centers | Number of<br>server racks | Fault tolerance<br>level |
> | --- | --- | --- | --- | --- | --- | --- | --- |
> | `none` | 1 | 1 | Node | Node | 1 | 1 | No fault tolerance |
> | `block-4-2` | 1.5 | 8 (10 recommended) | Rack | Data center | 1 | 8 | Can stand a failure of 2 racks |
> | `mirror-3-dc` | 3 | 9 (12 recommended) | Rack | Data center | 3 | 3 in each data center | Can stand a failure of a data center and 1 rack in one of the two other data centers. |
> | `block-4-2`<br>(reduced) | 1.5 | 10 | ½ a rack | Data center | 1 | 5 | Can stand a failure of 1 rack. |
> | `mirror-3-dc`<br>(reduced) | 3 | 12 | ½ a rack | Data center | 3 | 6 | Can stand a failure of a data center and 1 server in one of the two other data centers. |
> | `mirror-3-dc`<br>(3 nodes) | 3 | 3 | Server | Data center | 3 | Doesn't matter | Can stand a failure of a single server, or a failure of a data center. |

Reviewer: ```suggestion
The table below describes the requirements and fault tolerance levels for different operating modes:

| Mode | Storage<br>volume multiplier | Minimum<br>number<br>of nodes | Fail<br>domain | Fail<br>realm | Number of<br>data centers | Number of<br>server racks |
| --- | --- | --- | --- | --- | --- | --- |
| `none`, no fault tolerance | 1 | 1 | Node | Node | 1 | 1 |
| `block-4-2`, can stand a failure of 2 racks | 1.5 | 8 (10 recommended) | Rack | Data center | 1 | 8 |
| `mirror-3-dc`, can stand a failure of a data center and 1 rack in one of the remaining data centers | 3 | 9 (12 recommended) | Rack | Data center | 3 | 3 in each data center |
| `block-4-2` *(reduced)*, can stand a failure of 1 rack | 1.5 | 10 | ½ a rack | Data center | 1 | 5 |
| `mirror-3-dc` *(reduced)*, can stand a failure of a data center and 1 server in one of the two other data centers | 3 | 12 | ½ a rack | Data center | 3 | 6 |
| `mirror-3-dc` *(3 nodes)*, can stand a failure of a single server, or a failure of a data center | 3 | 3 | Server | Data center | 3 | Doesn't matter |
```

{% note info %}

> The storage volume multiplier specified above only applies to the fault tolerance factor. Other influencing factors (for example, slot fragmentation and granularity) must be taken into account for storage size planning.

Reviewer: ```suggestion
The storage volume multiplier specified above only applies to the fault tolerance factor. Other influencing factors (for example, [slot](./glossary.md#slot) fragmentation and granularity) must be taken into account for storage size planning.
```

{% endnote %}

> When creating a [storage group](glossary.md#storage-groups), which is a basic allocation unit for storage management, {{ ydb-short-name }} chooses [VDisks](./glossary.md#vdisk) that are located on [PDisks](./glossary.md#pdisk) from different fail domains. For `block-4-2` mode, a storage group should be distributed across at least 8 fail domains, and for `mirror-3-dc` mode, across 3 fail realms, with at least 3 fail domains in each realm.

Reviewer: ```suggestion
When creating a [storage group](glossary.md#storage-groups), which is a basic allocation unit for storage management, {{ ydb-short-name }} selects [VDisks](./glossary.md#vdisk) that are located on [PDisks](./glossary.md#pdisk) from different fail domains. For `block-4-2` mode, a storage group should be distributed across at least 8 fail domains, while for `mirror-3-dc` mode, it should be distributed across 3 fail realms, with at least 3 fail domains in each realm.
```

For information about how to set the {{ ydb-short-name }} cluster topology, see [{#T}](../deploy/configuration/config.md#domains-blob).

### Reduced configurations {#reduced}

> If it is impossible to use the [recommended amount](#cluster-config) of hardware, you can divide servers within a single rack into two dummy fail domains. In this configuration, the failure of one rack results in the failure of two domains instead of just one. In such reduced configurations, {{ ydb-short-name }} will continue to operate if two domains fail. The minimum number of racks in a cluster is five for `block-4-2` mode and two per data center (e.g. six in total) for `mirror-3-dc` mode.
> 
> The minimal fault-tolerant configuration of a {{ ydb-short-name }} cluster uses the `mirror-3-dc-3-nodes` operating mode, which requires 3 servers. Each server in such configuration acts as both a fail domain and a fail realm, and a cluster can stand a failure of a single server only.

Reviewer: ```suggestion
If it is impossible to use the [recommended amount](#cluster-config) of hardware, you can divide servers within a single rack into two dummy fail domains. In this configuration, the failure of one rack results in the failure of two domains instead of just one. In such reduced configurations, {{ ydb-short-name }} will continue to operate if two domains fail. The minimum number of racks in a cluster is five for `block-4-2` mode and two per data center (e.g., six in total) for `mirror-3-dc` mode.

The minimal fault-tolerant configuration of a {{ ydb-short-name }} cluster uses the `mirror-3-dc-3-nodes` operating mode, which requires only three servers. In this configuration, each server acts as both a fail domain and a fail realm, and the cluster can withstand the failure of only a single server.
```

## Fault tolerance {#fault-tolerance}

> If a disk fails (which may or may not include the failure of the server in which the disk is installed), {{ ydb-short-name }} can automatically reconfigure a storage group. In this reconfiguration, a new VDisk replaces the VDisk located on the failed hardware, and the system tries to place it on operational hardware during the reconfiguration process. The same rules apply as when creating a group: the new VDisk is created in a fail domain different from any other VDisks in the group (and within the same fail realm as the failed VDisk in the `mirror-3-dc` mode).
> 
> This setup can cause issues when a cluster's hardware is distributed across the minimum required number of fail domains:
> 
> * If an entire fail domain is down, reconfiguration becomes impractical, as a new VDisk can only be placed in the fail domain that is down.
> * Reconfiguration is possible only if part of a fail domain is down. However, the load previously handled by the failed hardware will be redistributed across the hardware, remaining in the same fail domain.
> 
> If the number of fail domains in a cluster exceeds the minimum amount required for creating storage groups at least by one (that is, 9 domains for `block-4-2` and 4 domains in each fail realm for `mirror-3-dc)`, in case some hardware fails, the load can be redistributed across all the hardware that is still running.

Reviewer: ```suggestion
If a disk fails, {{ ydb-short-name }} can automatically reconfigure a storage group. Whether the disk failure is caused by the whole server failure or not is irrelevant in this context. Such reconfiguration replaces the VDisk located on the failed hardware with a new VDisk, and the system tries to place it on operational hardware. The same rules apply as when creating a group:

* The new VDisk is created in a fail domain different from any other VDisks in the group.
* In the `mirror-3-dc` mode, it is created within the same fail realm as the failed VDisk.

This process can cause issues when a cluster's hardware is distributed across the minimum required number of fail domains:

* If an entire fail domain is down, reconfiguration becomes impractical, as a new VDisk can only be placed in the fail domain that is down.
* Reconfiguration is possible only if part of a fail domain is down. However, the load previously handled by the failed hardware will be redistributed across the hardware, remaining in the same fail domain.

The load can be redistributed across all the hardware that is still running if the number of fail domains in a cluster exceeds the minimum amount required for creating storage groups by at least one. This means having 9 domains for `block-4-2` and 4 domains in each fail realm for `mirror-3-dc`, which is recommended.
```

> The system can work with fail domains of any size. However, if there are few domains and a different number of disks in different domains, the number of storage groups that you can create will be limited. In this case, some hardware in fail domains that are too large may be underutilized. If the hardware is used in full, significant distortions in domain sizes may make reconfiguration impossible.
> 
> For example, there are 15 racks in a cluster with `block-4-2` mode. The first of the 15 racks hosts 20 servers and the other 14 racks host 10 servers each. To fully utilize all the 20 servers from the first rack, {{ ydb-short-name }} will create groups so that 1 disk from this largest fail domain is used in each group. As a result, if any other fail domain's hardware is down, the load can't be distributed to the hardware in the first rack.
> 
> {{ ydb-short-name }} can group disk drives of different vendors, capacities, and speeds. The resulting characteristics of a group depend on the set of the worst characteristics of the hardware serving the group. Generally, the best results can be achieved by using homogenous hardware. 
> 
> {% note info %}
> 
> Hardware from the same batch is more likely to have similar defects and may fail simultaneously. It is essential to consider this when building large-scale {{ ydb-short-name }} clusters.
> 
> {% endnote %}
> 
> Therefore, the optimal initial hardware configurations for production {{ ydb-short-name }} clusters look like this:
> 
> * **A cluster hosted in one availability zone**: This setup uses the `block-4-2` mode and consists of nine or more racks, each with an identical number of servers.
> * **A cluster hosted in three availability zones**: This setup uses the `mirror-3-dc` mode and is distributed across three data centers, with four or more racks in each and an identical number of servers.

Reviewer: ```suggestion
## Capacity and performance considerations

The system can work with fail domains of any size. However, if there are few domains and different numbers of disks in different domains, the number of storage groups that can be created will be limited. In this case, some hardware in fail domains that are too large may be underutilized. If all hardware is used fully, significant differences in domain sizes may make reconfiguration impossible.

For example, consider a cluster in `block-4-2` mode with 15 racks. The first of the 15 racks hosts 20 servers, while the other 14 racks host 10 servers each. To fully utilize all 20 servers from the first rack, {{ ydb-short-name }} will create groups so that 1 disk from this largest fail domain is used in each group. As a result, if the hardware in any other fail domain fails, the load cannot be redistributed to the hardware in the first rack.

{{ ydb-short-name }} can group disk drives of different vendors, capacities, and speeds. The resulting characteristics of a group depend on the set of the worst characteristics of the hardware serving the group. Generally, the best results can be achieved by using homogeneous hardware.

{% note info %}

Hardware from the same batch is more likely to have similar defects and may fail simultaneously. It is essential to consider this when building large-scale {{ ydb-short-name }} clusters.

{% endnote %}

Therefore, the optimal initial hardware configurations for production {{ ydb-short-name }} clusters are as follows:

* **A cluster hosted in one availability zone**: This setup uses the `block-4-2` mode and consists of nine or more racks, each with an identical number of servers.
* **A cluster hosted in three availability zones**: This setup uses the `mirror-3-dc` mode and is distributed across three data centers, with four or more racks in each, all containing an identical number of servers.
```

Auto reconfiguration of storage groups reduces the risk of data loss in the event of a sequence of failures, provided these failures occur with sufficient time intervals to recover redundancy. By default, reconfiguration begins one hour after {{ ydb-short-name }} detects a failure.

Once a group is reconfigured, a new VDisk is automatically populated with data to restore the required storage redundancy. This process increases the load on other VDisks in the group as well as on the network. The total data replication speed is limited on both the source and target VDisks to minimize the impact of redundancy recovery on system performance.

The time required to restore redundancy depends on the amount of data and hardware performance. For example, replication on fast NVMe SSDs may take an hour, while it could take more than 24 hours on large HDDs. To ensure reconfiguration is possible, a cluster should have free slots available for creating VDisks in different fail domains. When determining the number of slots to keep free, consider the risk of hardware failure, the time required to replicate data, and the time needed to replace failed hardware.

## Database availability

> A [database](./glossary.md#database) within a {{ ydb-short-name }} cluster is available if both its storage and compute is operational:
> 
> * all [storage groups](./glossary.md#storage-group) allocated for the database should be operational, e.g. stay within the allowed level of failures;
> * the compute resources of the currently available [database nodes](./glossary.md#database-node) (primarily, the amount of main memory) should be sufficient to start all the user objects (tables, topics) within the database, and to handle the user sessions.
> 
> To survive a data center failure at the database level, assuming a cluster configured for `mirror-3-dc` operating mode, the database nodes have to be running in all 3 data centers, and include sufficient resources to handle the full workload when running in just 2 of 3 data centers. This means having at least `35%` extra resources in terms of CPU and main memory when running on top of 3 fully available data centers.

Reviewer: ```suggestion
A [database](./glossary.md#database) within a {{ ydb-short-name }} cluster is available if both its storage and compute resources are operational:

- All [storage groups](./glossary.md#storage-group) allocated for the database must be operational, i.e., stay within the allowed level of failures.
- The compute resources of the currently available [database nodes](./glossary.md#database-node) (primarily the amount of main memory) must be sufficient to start all the [tablets](glossary.md#tablet) managing objects like [tables](glossary.md#table) or [topics](glossary.md#topic) within the database and to handle user sessions.

To survive an entire data center outage at the database level, assuming a cluster configured with the `mirror-3-dc` operating mode, the [database nodes](./glossary.md#database-node) must be evenly distributed between all 3 data centers and include sufficient resources to handle the entire workload when running in just 2 of the 3 data centers. To achieve this, database nodes in each datacenter need at least 35% extra spare CPU and RAM resources when running normally without ongoing failures. If database nodes are typically utilized above this threshold, consider adding more of them or moving them to servers with more resources.
```

{% endnote %}

For information about how to set the {{ ydb-short-name }} cluster topology, see [{#T}](../deploy/configuration/config.md#domains-blob).

## Production configurations of data storage

To ensure the required fault tolerance of {{ ydb-short-name }}, configure the [cluster disk subsystem](cluster/distributed_storage.md) properly: select the appropriate [fault tolerance mode](#fault-tolerance) and [cluster configuration](#cluster-config) for your cluster.

### Fault tolerance modes {#fault-tolerance}

We recommend using the following [fault tolerance modes](topology.md) for {{ ydb-short-name }} production installations:

* `block-4-2`: For a cluster hosted in a single availability zone.

> * `mirror-3-dc`: For a cluster hosted in three availability zones.

Author: Dropped that thing, duplicate indeed

A fail model of {{ ydb-short-name }} is based on concepts such as fail domain and fail realm.

Fail domain

: A set of hardware that may fail concurrently.

  For example, a fail domain includes disks of the same server (as all server disks may be unavailable if the server PSU or network controller is down). A fail domain also includes servers located in the same server rack (as the entire hardware in the rack may be unavailable if there is a power outage or some issue with the network hardware in the same rack).

  Any domain fail is handled automatically, without shutting down the system.

Fail realm

: A set of fail domains that may fail concurrently.

  An example of a fail realm is hardware located in the same data center that may fail as a result of a natural disaster.

> Usually a fail domain is a server rack, while a fail realm is a data center.

Author: Updated glossary.md, included in the PR

> When creating a [storage group](glossary.md#storage-groups), {{ ydb-short-name }} groups VDisks that are located on PDisks from different fail domains. For `block-4-2` mode, a PDisk should be distributed across at least 8 fail domains, and for `mirror-3-dc` mode, across 3 fail realms with at least 3 fail domains in each of them.

Author: Looks correct to me

This causes some issues when a cluster's hardware is distributed across the minimum required amount of fail domains:

* If the entire fail domain is down, reconfiguration no longer makes sense, since a new VDisk can only be located in the fail domain that is down.
* If part of a fail domain is down, reconfiguration is possible, but the load that was previously handled by the failed hardware will only be redistributed across hardware in the same fail domain.

If the number of fail domains in a cluster exceeds the minimum amount required for creating storage groups at least by one (that is, 9 domains for `block-4-2` and 4 domains in each fail realm for `mirror-3-dc)`, in case some hardware fails, the load can be redistributed across all the hardware that is still running.

The system can work with fail domains of any size. However, if there are few domains and a different number of disks in different domains, the number of storage groups that you can create will be limited. In this case, some hardware in fail domains that are too large may be underutilized. If the hardware is used in full, significant distortions in domain sizes may make reconfiguration impossible.

> For example, there are 15 racks in a cluster with `block-4-2` fault tolerance mode. The first of the 15 racks hosts 20 servers and the other 14 racks host 10 servers each. To fully utilize all the 20 servers from the first rack, {{ ydb-short-name }} will create groups so that 1 disk from this largest fail domain is used in each group. As a result, if any other fail domain's hardware is down, the load can't be distributed to the hardware in the first rack.

Author: Applied changes manually into the current version.

{{ ydb-short-name }} can group disks of different vendors, capacity, and speed. The resulting characteristics of a group depend on a set of the worst characteristics of the hardware that is serving the group. Usually the best results can be achieved if you use same-type hardware. When creating large clusters, keep in mind that hardware from the same batch is more likely to have the same defect and fail simultaneously.

Therefore, we recommend the following optimal hardware configurations for production installations:

* **A cluster hosted in 1 availability zone**: It uses `block4-2` fault tolerance mode and consists of 9 or more racks with the same amount of identical servers in each rack.
* **A cluster hosted in 3 availability zones**: It uses `mirror3-dc` fault tolerance mode and is distributed across 3 data centers with 4 or more racks in each of them, the racks being equipped with the same amount of identical servers.

See also [{#T}](#reduced).

> ### Redundancy recovery {#rebuild}
> 
> Auto reconfiguration of storage groups reduces the risk of data loss in the event of multiple failures that occur within intervals sufficient to recover the redundancy. By default, reconfiguration is done one hour after {{ ydb-short-name }} detects a failure.
> 
> Once a group is reconfigured, a new VDisk is automatically populated with data to restore the required storage redundancy in the group. This increases the load on other VDisks in the group and the network. To reduce the impact of redundancy recovery on the system performance, the total data replication speed is limited both on the source and target VDisks.
> 
> The time it takes to restore the redundancy depends on the amount of data and hardware performance. For example, replication on fast NVMe SSDs may take an hour, while on large HDDs more than 24 hours. To make reconfiguration possible in general, a cluster should have free slots for creating VDisks in different fail domains. When determining the number of slots to be kept free, factor in the risk of hardware failure, the time it takes to replicate data and replace the failed hardware.

Reviewer: ```suggestion
### Redundancy recovery {#rebuild}

Auto reconfiguration of storage groups reduces the risk of data loss in the event of a sequence of failures, provided these failures occur with sufficient time intervals to recover redundancy. By default, reconfiguration begins one hour after {{ ydb-short-name }} detects a failure.

Once a group is reconfigured, a new VDisk is automatically populated with data to restore the required storage redundancy. This process increases the load on other VDisks in the group as well as on the network. The total data replication speed is limited on both the source and target VDisks to minimize the impact of redundancy recovery on system performance.

The time required to restore redundancy depends on the amount of data and hardware performance. For example, replication on fast NVMe SSDs may take an hour, while it could take more than 24 hours on large HDDs. To ensure reconfiguration is possible, a cluster should have free slots available for creating VDisks in different fail domains. When determining the number of slots to keep free, consider the risk of hardware failure, the time required to replicate data, and the time needed to replace failed hardware.
```

> ### Simplified hardware configurations {#reduced}
> 
> If it's not possible to use the [recommended amount](#cluster-config) of hardware, you can divide servers within a single rack into two dummy fail domains. In this configuration, a failure of 1 rack means a failure of 2 domains and not a single one. In [both fault tolerance modes](#fault-tolerance), {{ ydb-short-name }} will keep running if 2 domains fail. If you use the configuration with dummy fail domains, the minimum number of racks in a cluster is 5 for `block-4-2` mode and 2 in each data center for `mirror-3-dc` mode.

Reviewer: ```suggestion
### Simplified hardware configurations {#reduced}

If it is impossible to use the [recommended amount](#cluster-config) of hardware, you can divide servers within a single rack into two dummy fail domains. In this configuration, the failure of one rack results in the failure of two domains instead of just one. In [both fault tolerance modes](#fault-tolerance), {{ ydb-short-name }} will continue to operate if two domains fail. When using the configuration with dummy fail domains, the minimum number of racks in a cluster is five for `block-4-2` mode and two per data center for `mirror-3-dc` mode.
```

### Fault tolerance level {#reliability}

The table below describes fault tolerance levels for different fault tolerance modes and hardware configurations of a {{ ydb-short-name }} cluster:

Fault tolerance<br>mode | Fail<br>domain | Fail<br>realm | Number of<br>data centers | Number of<br>server racks | Fault tolerance<br>level
:--- | :---: | :---: | :---: | :---: | :---
`block-4-2` | Rack | Data center | 1 | 9 or more | Can stand a failure of 2 racks
`block-4-2` | ½ a rack | Data center | 1 | 5 or more | Can stand a failure of 1 rack
`block-4-2` | Server | Data center | 1 | Doesn't matter | Can stand a failure of 2 servers
`mirror-3-dc` | Rack | Data center | 3 | 4 in each data center | Can stand a failure of a data center and 1 rack in one of the two other data centers

> `mirror-3-dc` | Server | Data center | 3 | Doesn't matter | Can stand a failure of a data center and 1 server in one of the two other data centers

Reviewer: This instance of the topologies list has some extra information, but I'd still merge it with the one in the beginning to avoid duplication and discrepancies between copies.

Author: Merged and revised the tables. Should be checked carefully, as I could get things wrong.

Node broker gives names to dynamic nodes when they register. By default, a node name consists of the hostname and the port on which the node is running.

In a dynamic environment where hostnames often change, such as in Kubernetes, using hostname and port leads to an uncontrollable increase in the number of unique node names. This is true even for a database with a handful of dynamic nodes. Such behavior may be undesirable for a time series monitoring system as the number of metrics grows uncontrollably. To solve this problem, the system administrator can set up *stable* node names.

A stable name identifies a node within the tenant. It consists of a prefix and a node's sequential number within its tenant. If a dynamic node has been shut down, after a timeout, its stable name can be taken by a new dynamic node serving the same tenant.

To enable stable node names, you need to add the following to the cluster configuration:

```yaml
feature_flags:
  enable_stable_node_names: true
```

By default, the prefix is `slot-`. To override the prefix, add the following to the cluster configuration:

```yaml
node_broker_config:
  stable_node_name_prefix: <new prefix>

> ```

Reviewer: This looks unrelated to distributed storage and topologies. I'd keep it where it was for now.

Author: Sure, fixed that (that was an error, moved NB config back to config)

> ## Sample cluster configurations {#examples}
> 
> You can find model cluster configurations for deployment in the [repository](https://github.com/ydb-platform/ydb/tree/main/ydb/deploy/yaml_config_examples/). Check them out before deploying a cluster.

Reviewer: ```suggestion
## See also

* [Documentation for DevOps Engineers](../devops/index.md)
* [Example configurations](https://github.com/ydb-platform/ydb/tree/main/ydb/deploy/yaml_config_examples/)
```
